wavedump/src/fit-histograms

```diff
@@ -0,0 +1,566 @@
+#!/usr/bin/env python3
+from __future__ import print_function, division
+import h5py
+import numpy as np
+import sys
+import csv
+import os
+import ROOT
+from ROOT import gROOT
+from ROOT import TMath
+
+# ROOT.ROOT.EnableThreadSafety()
+
+# DEFAULT VALUES FOR SPE FIT:
+D_OFFSET = 0
+D_LAMBDA = 1
+D_SPE_CHARGE = 0.8
+D_NOISE_SPREAD = 0.01
+D_SPE_CHARGE_SPREAD = 0
+D_ZERO_PEAK_SPREAD = 0.4
+# Number of peaks we use for SPE fitting
+n = 20
+
+def iqr(h):
+    """
+    Returns the iqr of the histogram `h`
+    """
+    l = None
+    r = None
+    total = h.GetEntries()
+    i = 0
+    n = h.GetBinContent(i)
+    while n / total < 0.25:
+        i += 1
+        n += h.GetBinContent(i)
+    high_prob = n / total
+    n -= h.GetBinContent(i)
+    i -= 1
+    low_prob = n / total
+    l = (((0.25 - low_prob) / (high_prob - low_prob)) * h.GetBinWidth(i)) + h.GetBinLowEdge(i)
+    while n / total < 0.75:
+        i += 1
+        n += h.GetBinContent(i)
+    high_prob = n / total
+    n -= h.GetBinContent(i)
+    i -= 1
+    low_prob = n / total
+    r = (((0.75 - low_prob) / (high_prob - low_prob)) * h.GetBinWidth(i)) + h.GetBinLowEdge(i)
+    return r-l
+
+def average_distance_from(val, h):
+    """
+    Standard deviation calculation, except use `val` instead of the mean.
+    """
+    d2 = 0
+    for i in range(h.GetMaximumBin() + 2):
+        d2 += (val - h.GetXaxis().GetBinCenter(i))**2 * h.GetBinContent(i)
+    return np.sqrt(d2)
+
+def save_as_csv(data_type, value, value_err, channel):
+    """
+    Saves charge data in `temp_data.csv`.
+    `temp_data.csv` is read by `save_spe_data.py`, which
+    organizes the data and saves it to `Fit_Data.csv`.
+    """
+    with open ('temp_data.csv', 'a') as file:
+        headers = [
+            'Data Type',
+            'Channel',
+            'Charge',
+            'Err Abs'
+            ]
+        writer = csv.DictWriter(file, headers)
+        if os.stat('temp_data.csv').st_size == 0:
+            writer.writeheader()
+        writer.writerow({
+            'Data Type': data_type,
+            'Channel': channel,
+            'Charge': value,
+            'Err Abs': value_err
+        })
```

Comment:
Think it would be cool if this function took as an argument the file name to save to. Probably also worth mentioning in the comment that it will append to the csv file if it already exists.

Reply:
I could do that for sure, but the file it saves to is meant to be purely temporary. Since the 511 histogram is fit separately from the SPE histogram, I needed a place to store the values until both were calculated. `save_spe_data.py` reads from the temporary file, and combines the 511 and SPE charge into one row on a more fleshed out csv file. I recently added an argument to `save_spe_data.py`to choose the file it writes to.

Reply:
Ah I understand now that eventually we want to have analyze and fit integrated, and possibly have the database do the light output calculation. I'll add this argument.